core/stage.go

```diff
@@ -77,6 +77,9 @@ type (
 
 		// Update persists an updated stage to the datastore.
 		Update(context.Context, *Stage) error
+
+		// Purge removes orpaned stages from the datastore.
+		Purge(context.Context) error
```

Comment:
Instead of having an explicit Stage purge function, I would handle all the cascading purges in the [Build.Purge](https://github.com/harness/drone/blob/master/store/build/build.go#L414) function.  This way, if we ever decide to switch to `ON DELETE` constraints, all code changes are isolated to the database code.